### PR TITLE
Make exercises page faster for power users

### PR DESCRIPTION
**Pull request by iHiD**
I have run these in production.

Screenshot 2023-04-27 at 02 00 16

### CODE REVIEW

File: db/schema.rb [hunk 1]
```diff
@@ -765,13 +765,13 @@
     t.integer "published_iteration_head_tests_status", default: 0, null: false
     t.integer "latest_iteration_head_tests_status", limit: 1, default: 0, null: false
     t.boolean "unlocked_help", default: false, null: false
-    t.index ["exercise_id", "published_at"], name: "index_solutions_on_exercise_id_and_published_at"
+    t.index ["exercise_id", "status", "num_stars"], name: "solutions_ex_stat_stars", order: { status: :desc, num_stars: :desc }
```
Comment (iHiD) [inline db/schema.rb]:
```suggestion
    t.index ["exercise_id", "published_at"], name: "index_solutions_on_exercise_id_and_published_at"
```

File: db/schema.rb [hunk 2]
```diff
@@ -788,6 +788,16 @@
...
+  create_table "submission_ai_help_records", charset: "utf8mb4", collation: "utf8mb4_unicode_ci", force: :cascade do |t|
+    t.bigint "submission_id", null: false
+    t.string "source", null: false
+    t.text "advice_markdown", null: false
+    t.text "advice_html", null: false
+    t.datetime "created_at", null: false
+    t.datetime "updated_at", null: false
+    t.index ["submission_id"], name: "fk_rails_76b9473637"
+  end
+
```
Comment (iHiD) [inline db/schema.rb]:
```suggestion
```

File: db/schema.rb [hunk 3]
```diff
@@ -1291,6 +1302,7 @@
   add_foreign_key "solutions", "exercises"
   add_foreign_key "solutions", "iterations", column: "published_iteration_id"
   add_foreign_key "solutions", "users"
+  add_foreign_key "submission_ai_help_records", "submissions"
```
Comment (iHiD) [inline db/schema.rb]:
```suggestion
```

File: db/schema.rb [hunk 4]
```diff
@@ -1207,6 +1217,7 @@
     t.index ["last_visited_on"], name: "index_users_on_last_visited_on"
     t.index ["paypal_payer_id"], name: "index_users_on_paypal_payer_id", unique: true
     t.index ["provider", "uid"], name: "index_users_on_provider_and_uid", unique: true
+    t.index ["reputation"], name: "index_users_on_reputation"
```
Comment (iHiD) [inline db/schema.rb]:
```suggestion
```

File: db/schema.rb [hunk 5]
```diff
@@ -1118,8 +1128,8 @@
     t.index ["earned_on"], name: "sweeper"
     t.index ["exercise_id"], name: "index_user_reputation_tokens_on_exercise_id"
     t.index ["track_id", "category", "external_url"], name: "index_user_reputation_tokens_on_track_id_category_external_url"
-    t.index ["track_id"], name: "index_user_reputation_tokens_on_track_id"
     t.index ["uniqueness_key", "user_id"], name: "index_user_reputation_tokens_on_uniqueness_key_and_user_id", unique: true
+    t.index ["user_id", "category"], name: "index_user_reputation_tokens_on_user_id_and_category"
```
Comment (iHiD) [inline db/schema.rb]:
```suggestion
```

File: db/schema.rb [hunk 6]
```diff
@@ -1118,8 +1128,8 @@
     t.index ["earned_on"], name: "sweeper"
     t.index ["exercise_id"], name: "index_user_reputation_tokens_on_exercise_id"
     t.index ["track_id", "category", "external_url"], name: "index_user_reputation_tokens_on_track_id_category_external_url"
-    t.index ["track_id"], name: "index_user_reputation_tokens_on_track_id"
     t.index ["uniqueness_key", "user_id"], name: "index_user_reputation_tokens_on_uniqueness_key_and_user_id", unique: true
```
Comment (iHiD) [inline db/schema.rb:1122]:
```suggestion
    t.index ["track_id"], name: "index_user_reputation_tokens_on_track_id"
    t.index ["uniqueness_key", "user_id"], name: "index_user_reputation_tokens_on_uniqueness_key_and_user_id", unique: true
```

File: app/assemblers/assemble_exercise_list.rb [hunk 7]
```diff
@@ -19,6 +19,9 @@ def exercises
   def solutions
     return [] unless user
 
+    # TODO: It's *MUCH* quicker to get all the exercises for the user and the track
+    # and then to filter out the ones we don't want, than it is to use IN (...) with ids.
+    # So refactor this to do that. And check for other places too.
```
Comment (ErikSchierboom) [inline app/assemblers/assemble_exercise_list.rb]: Something for later!

**Review by ErikSchierboom — APPROVED** [review]

**Pull request merged**